w3af/plugins/audit/open_api_auth.py

```diff
@@ -0,0 +1,372 @@
+        Currently the plugin expects only 401.
+        We can introduce 'expected_code_regex' configuration parameter
+        which set a regex for expected response codes.
+      * If the plugin is updated to send other HTTP methods (see above),
+        we may want to introduce another configuration parameter
+        to set expected response codes in this case.
+      * Severity of identified issues may depend on response codes
+        returned by the server.
+
+    :author: Artem Smotrakov (artem.smotrakov@gmail.com)
+    :author: Andres Riancho (andres.riancho@gmail.com)
+    """
+
+    def __init__(self):
+        AuditPlugin.__init__(self)
+        self._spec = None
+        self._expected_codes = [401]
+
+    def audit(self, freq, orig_response, debugging_id):
+        """
+        Check if an API endpoint requires authentication according to its API spec.
+
+        :param freq: A FuzzableRequest to the API endpoint.
+        :param orig_response: The HTTP response associated with the fuzzable request.
+        :param debugging_id: A unique identifier for this call to audit().
+        """
+        # The check works only with API endpoints.
+        # crawl.open_api plugin has to be called before.
+        if not self._is_api_spec_available():
+            om.out.information("Could not find an API specification, skip")
+            return
+
+        # The API spec must define authentication mechanisms.
+        if not self._has_security_definitions_in_spec():
+            om.out.information("The API spec has no security definitions, skip")
+            return
+
+        # The check only runs if the API specification
+        # requires authentication for a endpoint.
+        if not self._has_auth(freq):
+            om.out.information("Request doesn't have auth info, skip")
+            return
+
+        # Remove auth info from the request.
+        freq_with_no_auth = self._remove_auth(freq)
+
+        # Send the request to the server, and check if access denied.
+        self._uri_opener.send_mutant(freq_with_no_auth,
+                                     callback=self._check_response_code,
+                                     debugging_id=debugging_id)
+
+    def _check_response_code(self, freq, response):
+        """
+        Check if an HTTP request contains an expected response code.
+
+        :param freq: A FuzzableRequest to the API endpoint.
+        :param response: The HTTP response associated with the fuzzable request.
+        """
+        if response.get_code() not in self._expected_codes:
+            desc = 'A %s request without auth info was send to %s. ' \
+                   'The server replied with an unexpected code (%d), expected %s' \
+                   % (freq.get_method(), freq.get_url(),
+                      response.get_code(), self._expected_codes)
+
+            # Should severity depend on response codes?
+            # For example, 2xx codes results to HIGH, but 4xx may be MEDIUM/LOW
```

Comment:
Yes, this is a good idea to remove false positives.

The way I recommend doing it is to add a new check below `if not self._has_auth(freq)`, like this:

```python
        if not self._response_is_successful(freq):
            om.out.debug("OpenAPI request {some request info here} did not return a successful response. Skipping open_api_auth check for this endpoint.")
```

Where `_response_is_successful` would send the request (with credentials) and check that the response code is 200, 201, or some other codes which are commonly used to indicate success.

Reply:
Do you mean adding `_response_is_successful` check between `_remove_auth` and `_remove_auth` and `_uri_opener.send_mutant` calls in `audit()` method? But wasn't this request already sent? Is the result already available in `orig_response`?

What would be a definition of a successful response? 200 may be not enough since a endpoint may return 201, 202, etc. In fact, response codes should be defined in the spec. In general, it would be a nice improvement. Do you think such a check is strictly necessary?

In fact, I meant another thing here. I am talking about assigning different severity levels what we report a vulnerability to the KB. It should be done if/after `_response_is_successful` check passed.